cmds/switch_root/switch_root.go

```diff
@@ -23,140 +24,201 @@ func usage() string {
 	return "switch_root [-h] [-V]\nswitch_root newroot init"
 }
 
-// littleDoctor recursively deletes everything at "path" that
-// is in the same file system as "fs".
-func littleDoctor(path string, fsType int64) error {
-	var pathFS syscall.Statfs_t
+// get device for file descriptor
+func getDev(fd int) (dev uint64, err error) {
+	var stat syscall.Stat_t
 
-	if err := syscall.Statfs(path, &pathFS); err != nil {
-		return err
+	if err := syscall.Fstat(fd, &stat); err != nil {
```

Comment:
I'd recommend a convenience function for fstat/fstatat that should cover all the cases in this PR:

```go
func fstatat(fd int, path string) (unix.Stat_t, error) {
  var s unix.Stat_t
  flags := unix.AT_SYMLINK_NOFOLLOW | unix.AT_EMPTY_PATH
  if err := unix.Fstatat(fd, path, &s, flags); err != nil {
    return unix.Stat_t{}, err
  }
  return s, nil
}

func getDev(fd int) (uint64, error) {
  s, err := fstatat(fd, "")
  if err != nil {
    return 0, err
  }
  return s.Dev, nil
}
```